drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,977 @@
+{
+	const struct can_nxp_s32_config *config = dev->config;
+
+	Canexcel_Ip_RxTxIRQHandler(config->instance);
+}
+
+static void can_nxp_s32_err_callback(const struct device *dev,
+					Canexcel_Ip_EventType eventType,
+					uint32 u32SysStatus,
+					const Canexcel_Ip_StateType *canexcelState)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	enum can_state state;
+	struct can_bus_err_cnt err_cnt;
+	void *cb_data = data->state_change_cb_data;
+	can_tx_callback_t function;
+	int alloc;
+	void *arg;
+
+	switch (eventType) {
+	case CANEXCEL_EVENT_TX_WARNING:
+		LOG_WRN("Tx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_RX_WARNING:
+		LOG_WRN("Rx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_BUSOFF:
+		LOG_ERR("Bus Off (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR:
+		LOG_ERR("Error Format Frames (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR_FD:
+		LOG_ERR("Error Data Phase (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_PASSIVE:
+		LOG_ERR("Error Passive (error 0x%x)", u32SysStatus);
+		break;
+	default:
+		break;
+	}
+
+	can_nxp_s32_get_state(dev, &state, &err_cnt);
+	if (data->state != state) {
+		data->state = state;
+		if (data->state_change_cb) {
+			data->state_change_cb(dev, state, err_cnt, cb_data);
+		}
+	}
+
+	if (state == CAN_STATE_BUS_OFF) {
+		/* Abort any pending TX frames in case of bus-off */
+		for (alloc = 0; alloc < CAN_NXP_S32_MAX_TX; alloc++) {
+			/* Copy callback function and argument before clearing bit */
+			function = data->tx_cbs[alloc].function;
+			arg = data->tx_cbs[alloc].arg;
+
+			if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+				if (can_nxp_s32_abort_msg(config,
+						ALLOC_IDX_TO_TXMB_IDX(alloc))) {
+					LOG_ERR("Can't abort message !");
+				};
+
+				function(dev, -ENETUNREACH, arg);
+				k_sem_give(&data->tx_allocs_sem);
+			}
+		}
+	}
+}
+
+static void nxp_s32_msg_data_to_zcan_frame(Canexcel_RxFdMsg msg_data,
+							struct can_frame *frame)
+{
+	if (!!(msg_data.Header.Id & CANXL_TX_HEADER_IDE_MASK)) {
+		frame->flags |= CAN_FRAME_IDE;
+	}
+
+	if (!!(msg_data.Header.Id & CANXL_TX_HEADER_RTR_MASK)) {
+		frame->flags |= CAN_FRAME_RTR;
+	}
+
+	if (!!(frame->flags & CAN_FRAME_IDE)) {
+		frame->id = (msg_data.Header.Id & CANXL_IP_ID_EXT_MASK);
+	} else {
+		frame->id = ((msg_data.Header.Id & CANXL_IP_ID_STD_MASK)
+						>> CANXL_IP_ID_STD_SHIFT);
+	}
+
+	frame->dlc = (msg_data.Header.Control & CANXL_TX_HEADER_DLC_MASK)
+						>> CANXL_TX_HEADER_DLC_SHIFT;
+
+	if (!!(msg_data.Header.Control & CANXL_TX_HEADER_FDF_MASK)) {
+		frame->flags |= CAN_FRAME_FDF;
+	}
+
+	if (!!(msg_data.Header.Control & CANXL_TX_HEADER_BRS_MASK)) {
+		frame->flags |= CAN_FRAME_BRS;
+	}
+
+	memcpy(frame->data, msg_data.data, can_dlc_to_bytes(frame->dlc));
+
+#ifdef CONFIG_CAN_RX_TIMESTAMP
+	frame->timestamp = msg_data.timeStampH << 32 | msg_data.timeStampL;
```

Comment:
`frame->timestamp` is 16 bit. Are you sure this is correct?

Reply:
I didn't pay attention to the variable type, we can extra variable to 64bit ?
This device driver supports timestamp 64bit